pkg/upgrade/upgrade.go

```diff
@@ -652,3 +662,718 @@ func cleanupDeprecatedKueueVAPB(ctx context.Context, cli client.Client) error {
+
+func getOdhDashboardConfig(ctx context.Context, cli client.Client, applicationNS string) (*unstructured.Unstructured, error) {
+	odhConfig := &unstructured.Unstructured{}
+	odhConfig.SetGroupVersionKind(gvk.OdhDashboardConfig)
+
+	// Try to get the OdhDashboardConfig
+	err := cli.Get(ctx, client.ObjectKey{Name: "odh-dashboard-config", Namespace: applicationNS}, odhConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return odhConfig, nil
+}
+
+func CalculateContainerResourceLimits(odhConfig *unstructured.Unstructured, sizeType string) (map[string]string, error) {
+	containerSizes, err := getContainerSizes(odhConfig, sizeType)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(containerSizes) == 0 {
+		return getDefaultResourceLimits(), nil
+	}
+
+	return calculateResourceLimitsFromSizes(containerSizes), nil
+}
+
+// getDefaultResourceLimits returns default resource limits when no container sizes are found.
+func getDefaultResourceLimits() map[string]string {
+	return map[string]string{
+		"maxMemory": "120Gi",
+		"minMemory": "8Gi",
+		"maxCpu":    "30",
+		"minCpu":    "1",
+	}
+}
+
+// calculateResourceLimitsFromSizes processes container sizes and calculates min/max resource limits.
+func calculateResourceLimitsFromSizes(containerSizes []ContainerSize) map[string]string {
+	limits := &resourceLimits{}
+
+	for _, size := range containerSizes {
+		limits.updateMemoryLimits(size.Resources.Requests.Memory, size.Resources.Limits.Memory)
+		limits.updateCpuLimits(size.Resources.Requests.Cpu, size.Resources.Limits.Cpu)
+	}
+
+	return limits.toMap()
+}
+
+// resourceLimits holds the calculated min/max resource values.
+type resourceLimits struct {
+	maxMemory, minMemory string
+	maxCpu, minCpu       string
+}
+
+// updateMemoryLimits updates memory limits based on request and limit values.
+func (r *resourceLimits) updateMemoryLimits(requestMemory, limitMemory string) {
+	candidateMax := selectMaxResource(requestMemory, limitMemory, CompareMemory)
+	if candidateMax != "" {
+		r.maxMemory = selectMaxFromCandidates(r.maxMemory, candidateMax, CompareMemory)
+	}
+
+	candidateMin := selectMinResource(requestMemory, limitMemory, CompareMemory)
+	if candidateMin != "" {
+		r.minMemory = selectMinFromCandidates(r.minMemory, candidateMin, CompareMemory)
+	}
+}
+
+// updateCpuLimits updates CPU limits based on request and limit values.
+func (r *resourceLimits) updateCpuLimits(requestCpu, limitCpu string) {
+	candidateMax := selectMaxResource(requestCpu, limitCpu, CompareCpu)
+	if candidateMax != "" {
+		r.maxCpu = selectMaxFromCandidates(r.maxCpu, candidateMax, CompareCpu)
+	}
+
+	candidateMin := selectMinResource(requestCpu, limitCpu, CompareCpu)
+	if candidateMin != "" {
+		r.minCpu = selectMinFromCandidates(r.minCpu, candidateMin, CompareCpu)
+	}
+}
+
+// toMap converts resourceLimits to a map, applying defaults for missing values.
+func (r *resourceLimits) toMap() map[string]string {
+	if r.minMemory == "" {
+		r.minMemory = "1Mi"
```

Comment:
nit: not a big deal, but the `toMap` method should not have side effect, so it should not change any struct data field